### DISCUSSION TITLE
Deform the Visual Model of an ElasticMaterialObject together with Force Fields

### DISCUSSION THREAD

**Discussion by qpding** [SofaDefrost/SoftRobots#139] [Help - Programming with SoftRobots] [answered]
Hi Sofa community,

I'm now working on a project which is very similar to the CableGripper tutorial. Everything works for now except for the model visualization part. Detailed descriptions are as follows.
### Purpose
Simulate a cable driven elastic end effector, and make the "Visual Model" (STL) bend together with the "Force Fields" (VTK) visualization model.
### Part of the code
```
from stlib3.scene import MainHeader
from stlib3.physics.deformable import ElasticMaterialObject
from stlib3.physics.constraints import FixedBox
from softrobots.actuators import PullingCable
from splib3.loaders import loadPointListFromFile
import Sofa.Core
import Sofa.constants.Key as Key
```
```
def createScene(rootNode):
    """
    Description for createScene: Test
    """
    MainHeader(rootNode, gravity=[0.0,-981.0,0.0])
    rootNode.addObject('RequiredPlugin', pluginName='SoftRobots')
    rootNode.VisualStyle.displayFlags = "showForceFields showBehaviorModels showInteractionForceFields"

    rootNode.addObject('FreeMotionAnimationLoop')
    rootNode.addObject('GenericConstraintSolver', tolerance=1e-3, maxIterations=50)
    
    end_effector = rootNode.addChild('end_effector')
    eobject = ElasticMaterialObject(end_effector,
               volumeMeshFileName=path+file_name+'0.vtu',
               poissonRatio=0.3,
               youngModulus=18000,
               totalMass=0.0005,
               surfaceColor=[0.0, 0.8, 0.7],
               surfaceMeshFileName=path+file_name+'.stl',
               rotation=[0.0, 0.0, 0.0],
               translation=[0.0, 0.0, 0.0])
    end_effector.addChild(eobject)

    FixedBox(eobject, doVisualization=True,
            atPositions=[22.0, 0.0, 0.0, 25.0, 5.0, 5.0])

    cable = PullingCable(eobject,
            name="PullingCable",
            cableGeometry=loadPointListFromFile(path+"cable.json"),
            pullPointLocation=[20.0, 0.0, 0.0],
            rotation=[0.0, 0.0, 0.0],
            translation=[0.0, 0.0, 0.0])

    end_effector.addObject(TipController(cable))

    # superimpose stl render over the vtu
    end_effector.addObject('MeshSTLLoader', name='mesh', filename=path+file_name+'.stl')
    end_effector.addObject('OglModel', src="@mesh", color="0.8 0.7 0.7 0.9")
    #  end_effector.addObject('BarycentricMapping')

    return rootNode
```
### OS & Version
Debian Testing, Python3.9, SOFA V21.12.00.
#### STL is not bending
When the cable is actuated, the object (blue-ish "Force Fields", VTK) bends as expected, but the "Visual Models"/"OglModel" (STL) is kept straight.
#### surfaceMeshFile is not visible
The argument "surfaceMeshFileName=path+file_name+'.stl'" of **_eobject_** doesn't seem to work. No extra visualization except for the blue-ish Force Fields would show up whether the **_surfaceMeshFileName_** is provided or not. 
This happens in the CableGripper tutorial (SofaPython3 version) too.

### Trials
I've tried the CableGripper tutorial of Python2 version, where the STL model bends with the VTK model perfectly. Compared codes of Py2 and Py3 versions but found no big differences. It feels like that I'm missing the part to define the mapping between fields but got no idea on how to do it.

Any hint or help is appreciated.

Quinn

**Comment by EulalieCoevoet** [comment]
Hi Quinn, 

It's hard to see what could be wrong without your mesh files. When you say it does not bend:
- do you mean the visual model (surface mesh) does not move at all? It that case do you have any warnings or errors showing up on your terminal?
- if the surface mesh mostly follows the mechanical model (volume mesh) but has some parts that do not bend, it may just mean that you don't have enough points on your surface mesh. The mesh can only bend at the location of the points. 

As for the "no extra visualization except for the blue-ish Force Fields", there is a tab on the runSofa application's window named "View" where you can select what to display. Have you checked that the "Visual Models" check box is actually checked?  

Hope this help.
Eulalie

**Reply by qpding** [reply]
Hi Eulalie,

I went through the source codes and found the problem:
When defining an **_ElasticMaterialObject_** in Python3, the argument **_surfaceColor_** should have type Vec4d, instead of Vec3d as in the Python2 version. The 4th number should be transparency if I didn't make it wrong. I believe feeding a 3d vector to **_surfaceColor_** of P3 version should trigger an error in [this line](https://github.com/SofaDefrost/STLIB/blob/469f693b9071ba1bfc478d2f75e6a606d06a3f86/python3/src/stlib3/visuals/VisualModel.py#L32), and cause the OglModel failure, but no info pops up in the terminal.

The problem comes from the _**VisualModel**_ class of STLIB after comparing the codes of both python versions. Not exactly an issue of either SoftRobot Plugin or STLIB Plugin, and it's part of my fault so I'll keep this thread here.

However, it might help to update the SoftRobot tutorial for less confusion, e.g. [this line](https://github.com/SofaDefrost/SoftRobots/blob/a101685f754488acd807421c08263ca1e9554516/docs/sofapython3/tutorials/CableGripper/details/finger.py#L39).

Quinn

**Reply by EulalieCoevoet** [reply]
Hi Quinn, 

Thanks for the heads up! I agree that we should get an error in such case. 
We could maybe open an issue in SofaPython3? 

Anyway I'll update the tutorial as you suggest. 

Eulalie

**Reply by qpding** [reply]
Hi Eulalie,

Agree on filing an issue on this. But should it be in SofaPython3 or STLIB? The problem does come from
`self.addObject('OglModel', name="OglModel", src="@loader", ...)`
but I don't know if SofaPython3 is responsible for handling this issue, cuz it actually happens when defining a VisualModel object in STLIB.

Quinn

**Reply by EulalieCoevoet** [reply]
Actually when you give a Vec3 to an `OglModel` it adds a one at the end, which is a good default behavior in that case, in my opinion. 

The problem is that there's no typing by default in python rigth. So if we want to raise an error or a warning we should implement it. I think it could be done in Sofa.Prefab (which comes from SofaPython3) since we give a type to the properties. For instance this code should raise a warning, or at least complete the list with the default value, instead of printing `[1,1,0,0]`:

```python
import Sofa

class EmptyPrefab(Sofa.Prefab):

    properties = [
        {'name':'myProperty', 'type':'Vec4d', 'help':'',  'default':[1.,1.,1.,1.]}
    ]

    def __init__(self, *args, **kwargs):
        Sofa.Prefab.__init__(self, *args, **kwargs)
        print(self.myProperty.value)


def createScene(root):
    myPrefab = EmptyPrefab(myProperty=[1,1])
    root.addChild(myPrefab)
```

**Reply by qpding** [reply]
I see. Will go open an issue in SofaPython3 on this.

Thanks for all the help and careful explanation. I'll close this discussion for now.

**Comment by qpding** [comment, marked as answer]
To get a deformable visual model of an ElasticMaterialObject, use a Vec4d _**surfaceColor**_, e.g. surfaceColor= [0.1, 0.2, 0.3, 1.0], when defining the object.

If a Vec3d is used, a zero will be concatenated to the end, which makes the OglModel transparent.
